### PR TITLE
HPCC-19970 In simplified expr use no_simplified operator

### PR DESCRIPTION
**Pull request by shamser**
Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by shamser** [comment]
@ghalliday Please can you review

File: ecl/hql/hqlcache.cpp [hunk 1]
```diff
@@ -355,29 +355,22 @@ static IHqlExpression * createSimplifiedDefinitionFromType(ITypeInfo * type, boo
...
 
-static IHqlExpression * createSimplifiedBodyDefinition(IHqlExpression * expr, bool implicitConstantType=false)
+static IHqlExpression * createSimplifiedBodyDefinition(IHqlExpression * expr)
 {
     if (expr->isFunction())
     {
         if (!expr->isFunctionDefinition())
             return nullptr;
-        OwnedHqlExpr newBody = createSimplifiedBodyDefinition(expr->queryChild(0)->queryBody());
+        LinkedHqlExpr newBody = createSimplifiedBodyDefinition(expr->queryChild(0)->queryBody());
```
Comment (ghalliday) [inline ecl/hql/hqlcache.cpp]: This will leak memory

File: ecl/hql/hqlgram2.cpp [hunk 2]
```diff
@@ -12369,6 +12370,16 @@ bool verifySimpifiedDefinition(IHqlExpression *origExpr, IHqlExpression *simplif
     OwnedHqlExpr parsed = parseDefinition(ecl, simplifiedDefinition->queryId(), errors);
     if (!parsed)
     {
+#ifdef DEBUG
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp]: Should be have a leading underscore: _DEBUG

File: ecl/hql/hqlgram2.cpp [hunk 3]
```diff
@@ -12380,12 +12391,16 @@ bool verifySimpifiedDefinition(IHqlExpression *origExpr, IHqlExpression *simplif
 
         if (!streq(t1, t2))
         {
+#ifdef DEBUG
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp]: same here and following

File: ecl/hql/hqlir.cpp [hunk 4]
```diff
@@ -2031,6 +2032,7 @@ id_t ExpressionIRPlayer::doProcessExpr(IHqlExpression * expr)
         {
             HqlExprArray scopeSymbols;
             expr->queryScope()->getSymbols(scopeSymbols);
+            scopeSymbols.sort(compareSymbolsByName);
```
Comment (ghalliday) [inline ecl/hql/hqlir.cpp]: This has already been included in another PR - should really be rebasd on top of the current master

**Review by ghalliday — COMMENTED** [review]
A few initial comments.  I'll run a regression test as well and see if I spot any other issues.

**Comment by ghalliday** [comment]
@shamser

**Comment by shamser** [comment]
@ghalliday I've fixed those issues mentioned in your comments. I added an additional change:  "Unable to expand constant expression when using cache..." is only shown if syntax checking is enabled AND caching is enabled

**Comment by ghalliday** [comment]
@shamser a few more comments from running the regression suite:

codes2 shows a new entry for SIMPLFIED in an error message.  If should be included in the simplify list for expression hqlgram2.cpp(11408)

eclcc -internal reports an error:
    Error: 'simplified' is missing from reservedWords.cpp

digraph.ecl fails since SIMPLIFIED is no a reserved keyword.  I think you need to use __SIMPLIFIED__ instead to ensure that there are no other false clashes.

**Comment by shamser** [comment]
@ghalliday

**Comment by ghalliday** [comment]
@shamser You also need the keyword from SIMPLIFIED to something that will not cause clashes with existing code e.g., \_\_SIMPLIFIED\_\_.

**Comment by shamser** [comment]
@ghalliday

**Comment by ghalliday** [comment]
@shamser looks ready.  Please squash and I will merge

**Pull request merged**